torchmetrics/functional/audio/sdr.py

```diff
@@ -0,0 +1,189 @@
+
+import torch
+
+from torchmetrics.utilities.imports import _FAST_BSS_EVAL_AVAILABLE, _TORCH_GREATER_EQUAL_1_8
+
+if _FAST_BSS_EVAL_AVAILABLE:
+    if _TORCH_GREATER_EQUAL_1_8:
+        from fast_bss_eval.torch.cgd import toeplitz_conjugate_gradient
+        from fast_bss_eval.torch.helpers import _normalize
+        from fast_bss_eval.torch.linalg import toeplitz
+        from fast_bss_eval.torch.metrics import compute_stats
+
+        solve = torch.linalg.solve
+    else:
+        import numpy
+        from fast_bss_eval.numpy.cgd import toeplitz_conjugate_gradient
+        from fast_bss_eval.numpy.helpers import _normalize
+        from fast_bss_eval.numpy.linalg import toeplitz
+        from fast_bss_eval.numpy.metrics import compute_stats
+
+        solve = numpy.linalg.solve
+else:
+    toeplitz = None
+    toeplitz_conjugate_gradient = None
+    compute_stats = None
+    _normalize = None
+
+from torch import Tensor
+
+from torchmetrics.utilities.checks import _check_same_shape
+
+
+def sdr(
+    preds: Tensor,
+    target: Tensor,
+    use_cg_iter: Optional[int] = None,
+    filter_length: int = 512,
+    zero_mean: bool = False,
+    load_diag: Optional[float] = None,
+) -> Tensor:
+    r"""Signal to Distortion Ratio (SDR) [1,2,3]
+
+    .. note::
+        1) using this metrics requires you to have ``fast-bss-eval`` install. Either install as ``pip install
+        torchmetrics[audio]`` or ``pip install fast-bss-eval``
+
+        2) preds and target need to have the same dtype, otherwise target will be converted to preds' dtype
+
+        3) when pytorch<1.8.0, numpy will be used to calculate this metric, which causes ``sdr`` non-differentiable
+
+    Args:
+        preds:
+            shape ``[..., time]``
+        target:
+            shape ``[..., time]``
+        use_cg_iter:
+            If provided, an iterative method is used to solve for the distortion
+            filter coefficients instead of direct Gaussian elimination.
+            This can speed up the computation of the metrics in case the filters
+            are long. Using a value of 10 here has been shown to provide
+            good accuracy in most cases and is sufficient when using this
+            loss to train neural separation networks.
+        filter_length:
+            The length of the distortion filter allowed
+        zero_mean:
+            When set to True, the mean of all signals is subtracted prior to computation of the metrics
+        load_diag:
+            If provided, this small value is added to the diagonal coefficients of
+            the system metrics when solving for the filter coefficients.
+            This can help stabilize the metric in the case where some of the reference
+            signals may sometimes be zero
+
+    Raises:
+        ValueError:
+            If ``fast-bss-eval`` package is not installed
+
+    Returns:
+        sdr value of shape ``[...]`` if compute_permutation is False, else ``[..., spk]``
```

Comment:
there is no `compute_permutation` flag
```suggestion
        sdr value of shape ``[...]``
```